{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; gene_type \"transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000223972.5\"; transcript_id \"EN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1           2      3      4  5  6  7  \\\n",
       "0  chr1  HAVANA        gene  11869  14409  .  +  .   \n",
       "1  chr1  HAVANA  transcript  11869  14409  .  +  .   \n",
       "2  chr1  HAVANA        exon  11869  12227  .  +  .   \n",
       "3  chr1  HAVANA        exon  12613  12721  .  +  .   \n",
       "4  chr1  HAVANA        exon  13221  14409  .  +  .   \n",
       "\n",
       "                                                   8  \n",
       "0  gene_id \"ENSG00000223972.5\"; gene_type \"transc...  \n",
       "1  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "2  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "3  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  \n",
       "4  gene_id \"ENSG00000223972.5\"; transcript_id \"EN...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"C:/Users/sunny/Desktop/ABC/gencode.v22.annotation.gtf\"\n",
    "gene_annotation = pd.read_csv(file_path, skiprows=5, header=None, sep= \"\\t\")\n",
    "gene_annotation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25.178478717803955\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "starttime = time.time()\n",
    "file_path = \"C:/Users/sunny/Desktop/ABC/gencode.v22.annotation.gtf\"\n",
    "gene_id_list = []\n",
    "gene_type_list = []\n",
    "gene_name_list = []\n",
    "f = open(file_path, 'r')\n",
    "for line in f.readlines():\n",
    "    if line.startswith(\"#\"):\n",
    "        continue\n",
    "    gene_id = re.search(r'gene_id \".+?\"', line).group()[11:-1]\n",
    "    gene_type = re.search(r'gene_type \".+?\"', line).group()[11:-1]\n",
    "    gene_name = re.search(r'gene_name \".+?\"', line).group()[11:-1]\n",
    "    gene_name_list.append(gene_name)\n",
    "    gene_id_list.append(gene_id)\n",
    "    gene_type_list.append(gene_type)\n",
    "gene_type_df = pd.DataFrame(gene_type_list, columns=[\"gene_type\"])\n",
    "gene_id_df = pd.DataFrame(gene_id_list, columns=[\"gene_id\"])\n",
    "gene_name_df = pd.DataFrame(gene_name_list, columns=[\"gene_name\"])\n",
    "df1 = pd.concat([gene_name_df, gene_type_df, gene_id_df], axis=1)\n",
    "endtime = time.time()\n",
    "print(endtime-starttime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                           gene_type          gene_id\n",
       "0   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5\n",
       "1   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5\n",
       "2   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5\n",
       "3   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5\n",
       "4   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2563671, 9)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_annotation.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = gene_annotation.drop([8],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df1, df2],axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                           gene_type          gene_id     0  \\\n",
       "0   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "1   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "2   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "3   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "4   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "\n",
       "        1           2      3      4  5  6  7  \n",
       "0  HAVANA        gene  11869  14409  .  +  .  \n",
       "1  HAVANA  transcript  11869  14409  .  +  .  \n",
       "2  HAVANA        exon  11869  12227  .  +  .  \n",
       "3  HAVANA        exon  12613  12721  .  +  .  \n",
       "4  HAVANA        exon  13221  14409  .  +  .  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>ensmbol_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                           gene_type          gene_id     0  \\\n",
       "0   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "1   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "2   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "3   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "4   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  chr1   \n",
       "\n",
       "        1           2      3      4  5  6  7       ensmbol_id  \n",
       "0  HAVANA        gene  11869  14409  .  +  .  SG00000223972.5  \n",
       "1  HAVANA  transcript  11869  14409  .  +  .  SG00000223972.5  \n",
       "2  HAVANA        exon  11869  12227  .  +  .  SG00000223972.5  \n",
       "3  HAVANA        exon  12613  12721  .  +  .  SG00000223972.5  \n",
       "4  HAVANA        exon  13221  14409  .  +  .  SG00000223972.5  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"ensmbol_id\"] = df[\"gene_id\"].str[:15]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.iloc[:, [0,1,2,5,11]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>2</th>\n",
       "      <th>ensmbol_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>transcript</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                           gene_type          gene_id           2  \\\n",
       "0   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5        gene   \n",
       "1   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5  transcript   \n",
       "2   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5        exon   \n",
       "3   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5        exon   \n",
       "4   DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5        exon   \n",
       "\n",
       "        ensmbol_id  \n",
       "0  SG00000223972.5  \n",
       "1  SG00000223972.5  \n",
       "2  SG00000223972.5  \n",
       "3  SG00000223972.5  \n",
       "4  SG00000223972.5  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns = ['gene_name', 'gene_type', 'gene_id', \"type\", 'ensmbol_id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>type</th>\n",
       "      <th>ensmbol_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>transcript</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000223972.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563666</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>SG00000210195.2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>SG00000210195.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563667</th>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>SG00000210195.2</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000210195.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563668</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563669</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "      <td>transcript</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563670</th>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "      <td>exon</td>\n",
       "      <td>SG00000210196.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2563671 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name                           gene_type          gene_id  \\\n",
       "0         DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5   \n",
       "1         DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5   \n",
       "2         DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5   \n",
       "3         DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5   \n",
       "4         DDX11L1  transcribed_unprocessed_pseudogene  SG00000223972.5   \n",
       "...           ...                                 ...              ...   \n",
       "2563666     MT-TT                             Mt_tRNA  SG00000210195.2   \n",
       "2563667     MT-TT                             Mt_tRNA  SG00000210195.2   \n",
       "2563668     MT-TP                             Mt_tRNA  SG00000210196.2   \n",
       "2563669     MT-TP                             Mt_tRNA  SG00000210196.2   \n",
       "2563670     MT-TP                             Mt_tRNA  SG00000210196.2   \n",
       "\n",
       "               type       ensmbol_id  \n",
       "0              gene  SG00000223972.5  \n",
       "1        transcript  SG00000223972.5  \n",
       "2              exon  SG00000223972.5  \n",
       "3              exon  SG00000223972.5  \n",
       "4              exon  SG00000223972.5  \n",
       "...             ...              ...  \n",
       "2563666  transcript  SG00000210195.2  \n",
       "2563667        exon  SG00000210195.2  \n",
       "2563668        gene  SG00000210196.2  \n",
       "2563669  transcript  SG00000210196.2  \n",
       "2563670        exon  SG00000210196.2  \n",
       "\n",
       "[2563671 rows x 5 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['transcribed_unprocessed_pseudogene', 'unprocessed_pseudogene',\n",
       "       'miRNA', 'lincRNA', 'protein_coding', 'processed_pseudogene',\n",
       "       'antisense', 'processed_transcript', 'snRNA', 'sense_intronic',\n",
       "       'misc_RNA', 'transcribed_processed_pseudogene', 'TEC', 'snoRNA',\n",
       "       'scaRNA', 'rRNA', 'unitary_pseudogene', '3prime_overlapping_ncrna',\n",
       "       'polymorphic_pseudogene', 'sense_overlapping', 'pseudogene',\n",
       "       'IG_V_pseudogene', 'transcribed_unitary_pseudogene', 'sRNA',\n",
       "       'IG_V_gene', 'IG_C_gene', 'IG_J_gene', 'ribozyme', 'vaultRNA',\n",
       "       'non_coding', 'TR_C_gene', 'TR_J_gene', 'TR_V_gene',\n",
       "       'TR_V_pseudogene', 'IG_C_pseudogene',\n",
       "       'translated_unprocessed_pseudogene', 'macro_lncRNA', 'TR_D_gene',\n",
       "       'TR_J_pseudogene', 'translated_processed_pseudogene',\n",
       "       'IG_J_pseudogene', 'IG_D_gene', 'Mt_tRNA', 'Mt_rRNA'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"gene_type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2563671, 5)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name               OR4F5\n",
       "gene_type      protein_coding\n",
       "gene_id       SG00000186092.4\n",
       "type                     gene\n",
       "ensmbol_id    SG00000186092.4\n",
       "Name: 53, dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df.loc[53]\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>type</th>\n",
       "      <th>ensmbol_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000186092.4</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000186092.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>FO538757.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000279928.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000279928.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>FO538757.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000279457.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000279457.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>335</th>\n",
       "      <td>OR4F29</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000278566.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000278566.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>OR4F16</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000273547.1</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000273547.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563625</th>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000212907.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000212907.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563632</th>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000198886.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000198886.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563646</th>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000198786.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000198786.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563652</th>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000198695.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000198695.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2563660</th>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SG00000198727.2</td>\n",
       "      <td>gene</td>\n",
       "      <td>SG00000198727.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19814 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name       gene_type          gene_id  type       ensmbol_id\n",
       "53            OR4F5  protein_coding  SG00000186092.4  gene  SG00000186092.4\n",
       "128      FO538757.3  protein_coding  SG00000279928.1  gene  SG00000279928.1\n",
       "140      FO538757.2  protein_coding  SG00000279457.2  gene  SG00000279457.2\n",
       "335          OR4F29  protein_coding  SG00000278566.1  gene  SG00000278566.1\n",
       "434          OR4F16  protein_coding  SG00000273547.1  gene  SG00000273547.1\n",
       "...             ...             ...              ...   ...              ...\n",
       "2563625     MT-ND4L  protein_coding  SG00000212907.2  gene  SG00000212907.2\n",
       "2563632      MT-ND4  protein_coding  SG00000198886.2  gene  SG00000198886.2\n",
       "2563646      MT-ND5  protein_coding  SG00000198786.2  gene  SG00000198786.2\n",
       "2563652      MT-ND6  protein_coding  SG00000198695.2  gene  SG00000198695.2\n",
       "2563660      MT-CYB  protein_coding  SG00000198727.2  gene  SG00000198727.2\n",
       "\n",
       "[19814 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df.loc[(df[\"gene_type\"] == \"protein_coding\") & (df[\"type\"] == \"gene\"),:]\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"C:/Users/sunny/Desktop/ABC/c.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_csv(\"C:/Users/sunny/Desktop/ABC/protein_coding_gene_annotation.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
